### ISSUE TITLE
Running afl-collect with ASAN binary doesn't work as expected

### ISSUE THREAD

**Issue by kirit1193** [rc0r/afl-utils#53]
While running afl-collect with two binaries, one compiled with -fsanitize=memory set and one without, the former always classifies all crashes as invalid.

This is possible due to a memory limitation issue (since ASAN requires -m none to be set). However, memory limit cannot be modified since the flag is used for minimizing crash file-names.

Sample output while running afl-collect with ASAN binary:

```
[*] Going to collect crash samples from '/some/dir'.
[*] Found 12 fuzzers, collecting crash samples.
[*] Successfully indexed 285 crash samples.
[!] Removed 285 invalid crash samples from index.
[!] Removed 0 timed out samples from index.
[*] Copying 0 samples into output directory...
```

With gdb+exploitable:

```
[*] Going to collect crash samples from '/some/dir'.
[*] Found 12 fuzzers, collecting crash samples.
[*] Successfully indexed 285 crash samples.
[!] Removed 284 invalid crash samples from index.
[!] Removed 1 timed out samples from index.
[*] Generating intermediate gdb+exploitable script '/some/dir/gdb_script.0' for 0 samples...
[*] Executing gdb+exploitable script 'gdb_script.0'...
*** GDB+EXPLOITABLE SCRIPT OUTPUT ***
*** ***************************** ***
[!] Removed 0 duplicate samples from index. Will continue with 0 remaining samples.
[!] Removed 0 uninteresting crash samples from index.
[*] Generating final gdb+exploitable script '/some/dir/gdb_script' for 0 samples...
[*] Copying 0 samples into output directory...
```

**Comment by rc0r** [comment]
`afl-collect` does not know anything about memory limits and it does not need to. In case you run your fuzzer with a memory limit other than `-m none` you might find crashes because of some OOM condition. When you run your target with such a crash sample outside of `afl-fuzz` there's quite a chance that you can't reproduce the crash. That's exactly what `afl-collect` does. It collects your samples, feeds them into your target and checks whether the target process exits abnormally.

In case all your crashes are classified as invalid by `afl-collect` you're certainly dealing with crashes that only surfaced because of the memory limit `afl-fuzz` uses. You could use `ulimit` with the memory limit set for `afl-fuzz` before you run `afl-collect`. That way you should be able to reproduce the crashes in your target and thus `afl-collect` will detect them as well.

HTH!

**Comment by kirit1193** [comment]
I believed afl-collect didn't need to deal with memory limits as well, but I'm fuzzing a ASAN compiled binary which is giving me a mmap related crash when I run it with a specific input. Running the non ASAN compiled binary gives me a SIGBUS.

I can run afl-tmin with the ASAN compiled binary and minimize the input file in crash mode without any problems. 

However, running afl-collect failed to extract this crash from the collection directory.

I haven't tried setting a custom ulimit yet though.

**Comment by rc0r** [comment]
> I can run afl-tmin with the ASAN compiled binary and minimize the input file in crash mode without any problems.

That's the case because `afl-tmin` honors memory limits I think. What happens if you run `afl-tmin -m none` in crash mode?

**Comment by kirit1193** [comment]
Yes I am able to minimize the crashing input only with `-m none` set. Should've clarified.
Running without `-m none` gives the following:

```
[+] Read 32 bytes from 'inputfile'.
[*] Performing dry run (mem limit = 50 MB, timeout = 1000 ms)...
[+] Program exits with a signal, minimizing in crash mode.
[*] Stage #0: One-time block normalization...
[+] Block normalization complete, 28 bytes replaced.
[*] --- Pass #1 ---
[*] Stage #1: Removing blocks of data...
    Block length = 2, remaining size = 32
[+] Block removal complete, 32 bytes deleted.
[!] WARNING: Down to zero bytes - check the command line and mem limit!
[*] Stage #2: Minimizing symbols (0 code points)...
[+] Symbol minimization finished, 0 symbols (0 bytes) replaced.
[*] Stage #3: Character minimization...
[+] Character minimization done, 0 bytes replaced.
[*] --- Pass #2 ---
[*] Stage #1: Removing blocks of data...
    Block length = 1, remaining size = 0
[+] Block removal complete, 0 bytes deleted.

     File size reduced by : 100.00% (to 0 bytes)
    Characters simplified : 2800.00%
     Number of execs done : 24
          Fruitless execs : path=0 crash=0 hang=0

[*] Writing output to 't'...
[+] We're done here. Have a nice day!
```

**Comment by rc0r** [comment]
Well, this seems like a bug in `afl-tmin`. Anyways, could you reproduce your crashes in `afl-collect` after `ulimit`ing your virtual memory?

**Comment by kirit1193** [comment]
Sorry for the late update, ulimiting to values from 10GB to 20TB didn't help. I'm still getting all crashes categorized as invalid when running afl-collect with an ASAN compiled binary.

**Comment by rc0r** [comment]
Could you provide the command lines for the invocations of:

a) your `afl-fuzz` run **with** ASAN binary
b) your `afl-fuzz` run **without** ASAN binary
c) your `afl-collect` run **with** ASAN binary
d) your `afl-collect` run **without** ASAN binary

Without some more details I can't do much to diagnose the problem.

**Comment by kirit1193** [comment]
Command lines:

- `afl-fuzz` run *with* ASAN binary:

`afl-fuzz -i input -o output -m none -- ./magick_asan convert @@ /dev/null`

Running without -m none doesn't allow fuzzing to start, as expected.

- `afl-fuzz` run *without* ASAN binary:

`afl-fuzz -i input -o output -m 250 -- ./magick convert @@ /dev/null`

I can remove the -m 250 and get the same results mostly

- `afl-collect` run *with* ASAN binary:

`afl-collect -m -r -rr -j8 -e gdb_script ./output ./sorted_crashes -- ./magick_asan convert @@ /dev/null`

This categorizes all crashes as invalid. There are around 260 of them

- `afl-collect` run *without* ASAN binary:

`afl-collect -m -r -rr -j8 -e gdb_script ./output ./sorted_crashes -- ./magick convert @@ /dev/null`

Unfortunately ImageMagick without ASAN isn't crashing with any of these input cases, so around 19-20 of the conditions are being categorized as timeouts and the rest are categorized as invalid.
